### DISCUSSION TITLE
Proposal of implementation for localStorage integration

### DISCUSSION THREAD

**Discussion by williamneto** [idom-team/idom#1083] [Ideas]
So, i was trying to include some login and user management features in a reactpy app and stumble uppon the fact that we dont have any way to access localStorage or any other client side stored data yet.

Thinking about that, i started implementing a way to integrate localStorage from the browser with the backend.
After some experimentation i was able to achieve this with the following logic:
- The backend will keep a copy of the localStorage content
- When loaded, the client will send to the server all items in its localStorage
- The server will receive the itens from localStorage and keep it in an object
- When in server side any item from the localStorage copy is changed, a message is sent to the client to update the real localStorage

The current state of this implementation is in this [branch](https://github.com/williamneto/reactpy/tree/local-storage) in my fork.
In `src/py/reactpy/reactpy/samples/use_local_storage.py` is a sample on how to use this method to access and alter localStorage.

How i wasnt yet able to make the login page using this new feature, i wont open a pull request for now.
But would like to bring to discussion and see what you guys think about this idea.

**Comment by rmorshea** [comment]
A while back added the ability for messages being sent between the client/server to have a ["type"](https://github.com/reactive-python/reactpy/blob/main/src/js/packages/%40reactpy/client/src/messages.ts). This is, at present, basically unused, but I think it's a perfect use case for it. One can image `local-storage-set` and `local-storage-get` message types which could be sent back and forth between the client/server. While there exists [some infrastructure client-side](https://github.com/reactive-python/reactpy/blob/main/src/js/packages/%40reactpy/client/src/reactpy-client.ts#L8-L24) for subscribing to certain message types, no such system exists server-side. Adding this capability is on my radar and something I'll be working on in the coming weeks.

The particular use case mentioned here is related to a broader discussion about how users should handle authentication in a ReactPy app (https://github.com/reactive-python/reactpy/issues/828). While, from [my understanding](https://www.youtube.com/watch?v=M6N7gEZ-IUQ), you should always protect yourself against XSS, regardless of how client-side tokens are stored, doing so in local storage can make it easier to exploit. As such, whatever auth solution ReactPy provides will likely rely on HTTP-ONLY cookies (see solution 4 in above issue).

**Reply by Archmonger** [reply]
For the purpose of auth specifically, I think it's worth considering circumventing the client side entirely.

Right now, any framework with auth support retains a copy of the active user's session on server-side. There isn't any reason for us to do a round-trip from server->client->server just to fetch that data.

It's healthier for us to develop an implementation on a per-framework basis for server-side auth access. For example, Django [`SessionMiddleware`](https://channels.readthedocs.io/en/stable/topics/sessions.html) stores the current Django auth session within the websocket scope. We could develop, or have dependencies for, ASGI middleware that does the same.

**Reply by williamneto** [reply]
Yeah, i wasn't considering the XSS risks until now, thanks @rmorshea . 

I had passed through issue #828 but wasn't able to reproduce the [idom-auth-example-sanic](https://github.com/phihos/idom-auth-example-sanic) suggestion. Because it uses the old nomenclature of the project i wasnt able go around and fix it. If there is any updates in this matter i wold be happy to contribute.

Anyway, even if not used for authentication purposes, maybe my use_local_storage method could be useful in the future.

**Reply by Archmonger** [reply]
As a general rule, you should [never trust anything coming from a client](https://medium.com/@berniedurfee/never-trust-a-client-not-even-your-own-2de342723674).

The only thing that can be trusted is the server. The client should only be used to store non-important (or [high entropy](https://crypto4a.com/2021/07/13/why-we-need-entropy-in-cybersecurity/#:~:text=In%20cryptography%2C%20entropy%20is%20used,security%20value%20of%20the%20key.)) data.

**Comment by Archmonger** [comment]
Although having `localStorage` access would be great for us, I do not believe `localStorage` access is required for an implementation of ReactPy auth sessions.

See my [other comment](https://github.com/reactive-python/reactpy/discussions/1083#discussioncomment-6338956).

**Comment by williamneto** [comment]
I was looking into issue #1075 and localStorage access is part of it. 

@Archmonger did you see how i did it? ([here](https://github.com/reactive-python/reactpy/commit/ca3002cb4aeaa6c7af3aab75afffa28e840d45a3)) Could easily apply the same logic for sessionStorage but would be good to have some feedback first.

Then i was thinking about proceeding with the other items listed  in #1075

**Reply by Archmonger** [reply]
I think it's going to be better to develop local storage a different way.

We need an interface that allows the user to "refresh" to see what the latest values are in the client.

A better architecture could have an interface that looks like this:

```python
# Note: `local_storage` relies on `use_query`, and `local_storage_updater` relies on `use_mutation`
local_storage, local_storage_updater = use_local_storage()

print(local_storage.loading)  # The client still hasn't sent back `localStorage` contents
print(local_storage.error)  # Error code if things failed on the client
print(local_storage.data)  # The `localStorage` data, if it has been fetched
local_storage.refetch()  # The user can call `refetch` again to repeat. Will only refetch if there's no ongoing fetch.

local_storage_updater.execute( {...} )  # The client gets queued to update `localStorage`
print(local_storage_updater.loading)  # The client is still updating `localStorage` contents
print(local_storage_updater.error)  # Error code if things failed on the client
local_storage_updater.execute( { ... } )  # The user can call `execute` again to repeat. Will only re-execute if there's no ongoing execution.
```

This design will likely rely on [`reactpy_django.use_query`](https://github.com/reactive-python/reactpy-django/blob/0aaaaac8c44a1eafe34d7be0fb79ad710a0ae2dc/src/reactpy_django/hooks.py#L101) and [`reactpy_django.use_mutation`](https://github.com/reactive-python/reactpy-django/blob/0aaaaac8c44a1eafe34d7be0fb79ad710a0ae2dc/src/reactpy_django/hooks.py#L218) to be ported to `reactpy` core.

This architecture would be duplicated for every client-side feature,

**Reply by Archmonger** [reply]
I've updated my comment above with an easier to develop approach.

**Reply by williamneto** [reply]
Seems like a little bit too much. the way i see would be better to integrate localStorage and other client side features in a more embedded way, as part of the whole application life cycle. Not something that needed to be fetched and refreshed

One of the great advantages of reactpy is to unify frontend and backend in one single layer for the developer and the need to activelly execute a call to refresh localStorage looks like a loss for me. Would be too much more natural to just call a localStorage object and get or set the item that you want (as you would do in the browser), and leave reactpy automaticaly handle the fetch and refreshing process with the whole app.

**Reply by rmorshea** [reply]
@williamneto the interface above tries to answer the question, "what should be rendered while waiting for the client to respond with the contents of `localStorage`?" If the answer is nothing, then you can likely ignore the other status information. If instead, you'd like to say, present a spinner to the user while waiting, then `local_storage.loading` would allow you to do that.

So, if you don't care about displaying something while waiting, a component might look something like:

```python
@component
def example():
    local_storage = use_local_storage()
    return html.p(f"Hello, {local_storage.data.get('name', 'default')}!")
```

In this situation, the component will render "Hello, default!" until the client responds with the contents of `localStorage`
or if `name` is not in `localStorage`. Once the client responds, the component will render "Hello, {name}!" if the client did have a value for `name` in `localStorage`.

```python
@component
def example():
    local_storage = use_local_storage()
    if local_storage.loading:
        return html.p("Loading...")
    client_local_storage = local_storage.data or {}
    return html.p(f"Hello, {client_local_storage.get('name', 'default')}!")
```

Here, the component will render "Loading..." until the client responds with the contents of `localStorage`. Once the client responds, the component will render "Hello, default!" if the client did not have a value for `name` in `localStorage`, or "Hello, {name}!" if the client did have a value for `name` in `localStorage`.

**Reply by williamneto** [reply]
So @rmorshea, i see two problems with this logic.

First i don't know if would be good to implement this anything different than what already exists in the client interface. As i see, the [localStorage](https://developer.mozilla.org/pt-BR/docs/Web/API/Window/localStorage) is a very consolidated component of web architecture, and to present anything different from the original  would just make it unpractical to any developer that had already used localStorage.

With that in mind, i don't believe that the question "what should be rendered while waiting for the client to respond with the contents of localStorage?" should be answered in the localStorage component. this is a upper level decision, and should be handled in some way like that:

```
@component
def exxample():
    storage = use_local_storage()
    user = storage.get_item('user') # return none if item doesnt exists
    
    return html.h1(
        f"Welcome {user}" if user else "Loading..."
    )
```

For localStorage and other client side features, i believe it would be more practical for us to develop and would make adoption easier if we stick to the original specifications of these components.

**Reply by Archmonger** [reply]
Unfortunately `f"Welcome {user}" if user else "Loading..."` does not work. Because this does not provide any context as to whether `user` is actually missing within `local_storage`, if it just hasn't loaded yet, or if for some reason the client had an exception when trying to load it.

**Reply by rmorshea** [reply]
Given how ReactPy works, accessing client-side `localStorage` will require a network call. This is fundamentally different to how `localStorage` works in client-side Javascript so there are limitation in how closely we can mirror the JS spec. With that said, I think I'd be in favor of a more primitive storage interface in which `get_item` and `set_item` are async methods. Higher level hooks could be implemented around those primitives.

**Reply by Archmonger** [reply]
I personally would rather have a finalized implementation over primitives, since we don't want to enable users to develop with bad programming patterns.

**Reply by rmorshea** [reply]
@Archmonger, We'd previously talked about having generic `use_query` and `use_mutation` hooks that could be used with any database so long as you brought along the bindings. Local storage is just another kind of database, so if we supplied the primitives in `reactpy`, then instead of copying the `use_query/use_mutation` pattern into bespoke hooks for local storage we could just create a binding for the generic hooks.

**Reply by Archmonger** [reply]
I like that approach. Falls neatly into place.

**Reply by williamneto** [reply]
Looks like we are talking about different approach's here.

@rmorshea of course we will need a network call, thats exactly what i did in my fork. i just don't think that this network call needs to be handled by the developer. for me, reactpy could really well handle all this background communication and leave it simple to developers.

And i don't think that will be a case for what you guys are thinking because the localStorage won't change by itself in the client. any changes in localStorage would come from the server and then propagated to the client. So i really dont see why this case of "if the data doesnt exist what i do" is useful at all. We seem to be taking an unnecessary detour to implement something that is essentially simple.

But will keep an eye on the project in the hope i can learn more with yours approach

**Reply by rmorshea** [reply]
Ok, I think I understand. In your implementation, you sync local storage with the server before the first render. At that point, we just assume that local storage is always controlled server-side.

**Reply by williamneto** [reply]
Yes, i thought that was the most practical way. 

We could even add a security layer in the future by natively encrypting the localStorage in the server side before sending it to the client. This would guarantee integrity and security of data stored there.

Do you think we could apply this logic to other features listed in #1075 ?

**Reply by Archmonger** [reply]
Unfortunately, sending it before first paint ultimately slows down the speed of first paint. This ends up being an exponentially larger problem as more clients connect to one server.

**Reply by rmorshea** [reply]
Leaving technical questions aside, here's what the two interfaces would look like:

```python
# SYNC USAGE
local_storage = use_local_storage()
value, set_value = use_state(local_storage.get_item("value"))

@use_effect
def set_local_storage_value():
    local_storage.set_item(value)
```

```python
# ASYNC USAGE
local_storage = use_local_storage()
value, set_value = use_state(None)

@use_effect
async def get_local_storage_value():
    set_value(await local_storage.get_item("value"))

@use_effect
async def set_local_storage_value():
    await local_storage.set_item(value)

if values is None:
    # value has not loaded yet
    ...
else:
    # value has loaded
    ...
```

Regardless of what approach you take `set_item` ought to be in an effect since it's a "side effect". So the main difference is that in the async API you need an extra effect to retrieve the value.

**Reply by williamneto** [reply]
@rmorshea yeah, its working like that.

Adapted this [sample](https://github.com/williamneto/reactpy/blob/web-application-apis/src/py/reactpy/reactpy/samples/use_local_storage.py) to the async usage.
